code/modules/reagents/reagent_containers/ampoules.dm

```diff
@@ -0,0 +1,402 @@
+		clean = FALSE
+	else if(HAS_TRAIT(user, TRAIT_CHUNKYFINGERS) && prob(30))
+		clean = FALSE
+	else if(iscarbon(user))
+		var/mob/living/carbon/carbon_user = user
+		if(prob(carbon_user.jitteriness))
+			clean = FALSE
+
+	if(clean)
+		snap()
+	else
+		obj_break()
+	return TRUE
+
+/obj/item/reagent_containers/glass/ampoule/welder_act(mob/living/user, obj/item/tool)
+	if(sealed || user.combat_mode)
+		return ..()
+	var/heat = tool.get_temperature()
+	if(heat)
+		reagents.expose_temperature(heat)
+	try_weld_shut(tool, user)
+	return TRUE
+
+/obj/item/reagent_containers/glass/ampoule/wirecutter_act(mob/living/user, obj/item/tool)
+	if(!sealed || user.combat_mode)
+		return ..()
+	var/heat = tool.get_temperature()
+	if(heat)
+		reagents.expose_temperature(heat, coeff = 0.01)
+	try_cut_open(tool, user)
+	return TRUE
+
+/obj/item/reagent_containers/glass/ampoule/attackby(obj/item/tool, mob/living/user, params)
+	if(!sealed || !tool.force || !tool.sharpness || user.combat_mode)
+		return ..()
+	var/heat = tool.get_temperature()
+	if(heat)
+		reagents.expose_temperature(heat, coeff = 0.01)
+	try_cut_open(tool, user)
+	return TRUE
+
+/obj/item/reagent_containers/glass/ampoule/try_splash(mob/user, atom/target)
+	if(sealed)
+		if(prob(20))
+			snap()
+		else
+			obj_break()
+	return ..()
+
+/obj/item/reagent_containers/glass/ampoule/process(delta_time)
+	var/datum/gas_mixture/env = loc?.return_air()
+	if(env)
+		air_contents.archive()
+		air_contents.temperature_share(env)
+
+	air_contents.react(src)
+	reagents.expose_temperature(air_contents.temperature)
+	if(air_contents.return_pressure() > TANK_LEAK_PRESSURE)
+		obj_break(BOMB)
+
+/**
+ * Populates the internal gas mixture of the ampoule.
+ */
+/obj/item/reagent_containers/glass/ampoule/proc/populate_gases(source = initial_gas_mix)
+	if(!source)
+		return FALSE
+
+	if(istext(source))
+		air_contents.parse_gas_string(source)
+	else if(istype(source, /datum/gas_mixture))
+		var/datum/gas_mixture/source_mix = source
+		air_contents.copy_from(source_mix, air_contents.volume / source_mix.volume)
+	else if(islist(source))
+		var/list/source_list = source
+		var/list/cached_gas = air_contents.gases
+		cached_gas &= source_list
+		for(var/gas_id in source_list)
+			if(gas_id == "TEMP")
+				air_contents.temperature = source_list[gas_id]
+				continue
+			ADD_GAS(gas_id, cached_gas)
+			cached_gas[gas_id][MOLES] = source_list[MOLES]
+	else
+		CRASH("Invalid source - [isnull(source) ? "NULL" : source] - passed to ampoule.populate_gas() .")
+
+	START_PROCESSING(SSobj, src)
+	return TRUE
+
+/**
+ * Seals this ampoule.
+ *
+ * Makes the ampoule incapable of being poured out/into.
+ * Internalizes some of the gases from the environment outside of the ampoule.
+ *
+ * Arguments:
+ * - assume_env: Whether the ampoule should bottle up some of the gases in the environment.
+ */
+/obj/item/reagent_containers/glass/ampoule/proc/seal(assume_env = TRUE)
+	sealed = TRUE
+	spillable = FALSE
```

Comment:
What's the difference between these two?

Reply:
One is on `/obj/item/reagent_containers` and the other is on the ampoule.

Reply:
hhhhhhhhhhh